x-pack/filebeat/input/s3/input.go

```diff
@@ -0,0 +1,327 @@
+	"github.com/elastic/beats/filebeat/util"
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+var (
+	// input name
+	inputName = "s3"
+	// MaxNumberOfMessage at one poll
+	MaxNumberOfMessage int64 = 10
+	// WaitTimeSecond for each poll
+	WaitTimeSecond int64 = 20
+)
+
+type s3Info struct {
+	name string
+	key  string
+}
+
+func init() {
+	err := input.Register(inputName, NewInput)
+	if err != nil {
+		panic(err)
+	}
+}
+
+// Input is a input for s3
+type Input struct {
+	started  bool
+	outlet   channel.Outleter
+	config   config
+	cfg      *common.Config
+	registry *harvester.Registry
+	logger   *logp.Logger
+}
+
+// NewInput creates a new s3 input
+func NewInput(cfg *common.Config, outletFactory channel.Connector, context input.Context) (input.Input, error) {
+	cfgwarn.Beta("s3 input type is used")
+
+	logger := logp.NewLogger(inputName)
+
+	config := defaultConfig
+	if err := cfg.Unpack(&config); err != nil {
+		return nil, errors.Wrap(err, "failed unpacking config")
+	}
+
+	awsConfig := defaults.Config()
+	awsCredentials := awssdk.Credentials{
+		AccessKeyID:     config.AccessKeyID,
+		SecretAccessKey: config.SecretAccessKey,
+	}
+
+	if config.SessionToken != "" {
+		awsCredentials.SessionToken = config.SessionToken
+	}
+
+	awsConfig.Credentials = awssdk.StaticCredentialsProvider{
+		Value: awsCredentials,
+	}
+
+	outlet, err := outletFactory(cfg, context.DynamicFields)
+	if err != nil {
+		return nil, err
+	}
+
+	p := &Input{
+		started:  false,
+		outlet:   outlet,
+		cfg:      cfg,
+		config:   config,
+		logger:   logger,
+		registry: harvester.NewRegistry(),
+	}
+
+	return p, nil
+}
+
+// Run runs the input
+func (p *Input) Run() {
+	p.logger.Debugf("s3", "Run s3 input with queueURLs: %+v", p.config.QueueURLs)
+	if len(p.config.QueueURLs) == 0 {
+		p.logger.Error("No sqs queueURLs configured")
+		return
+	}
+
+	awsConfig := defaults.Config()
+	awsCredentials := awssdk.Credentials{
+		AccessKeyID:     p.config.AccessKeyID,
+		SecretAccessKey: p.config.SecretAccessKey,
+	}
+	if p.config.SessionToken != "" {
+		awsCredentials.SessionToken = p.config.SessionToken
+	}
+
+	awsConfig.Credentials = awssdk.StaticCredentialsProvider{
+		Value: awsCredentials,
+	}
+
+	forwarder := harvester.NewForwarder(p.outlet)
+	for _, queueURL := range p.config.QueueURLs {
+		var d *util.Data
+		regionName, err := getRegionFromQueueURL(queueURL)
+		if err != nil {
+			p.logger.Errorf("failed to get region name from queueURL: %s", queueURL)
+			continue
+		}
+
+		awsConfig.Region = regionName
+		svcSQS := sqs.New(awsConfig)
+		svcS3 := s3.New(awsConfig)
+
+		// RECEIVE
+		receiveMessageInput := &sqs.ReceiveMessageInput{
+			QueueUrl:              &queueURL,
+			MessageAttributeNames: []string{"All"},
+			MaxNumberOfMessages:   &MaxNumberOfMessage,
+			VisibilityTimeout:     awssdk.Int64(20), // 20 seconds
```

Comment:
I guess this should be a const too, what is this timeout about?

Reply:
good point, I will update with some comments.